feat: реализовать механику чата тет а тет на консоли

beerbis opened the pull request.

Commit:
feat: реализовать механику чата тет а тет на консоли

Commit:
fix: учесть коментарии на "ревью", пусть это так называется

Commit:
chore: двинуть строки местами

Примечания
=========
Вооот наверно кто имелся в виду, когда говорилось про поля, что должны вверху быть.
Одно поле точно есть, да.. Я его воспринимал как топовый синглетон, коим по совести ему бы и быть, описание которого в классе и не обусловленно необходимостью, пусть будет был косяк.

beerbis commented on task1/src/Level2/Net/ChatingChannel.java:
```diff
@@ -0,0 +1,110 @@
+
+    public boolean isClosed() { return closed; }
+
+    public ChatingChannel(Socket socket, String myNickname, Consumer<String> logTheChat) throws IOException {
```
а вот что действительно интересно в плане обработки исключений - конструкторы кидающие исключения.
если из конструктора оно валится - надо ли в теле конструктора finally-ться на каждом шаге? а мало ли такая-то строка выкинет исключение, а ты ресурсы зажрал. Или на них всё равно потому что сборщик мусора?
пример:
`out = new DataOutputStream(socket.getOutputStream());` - пройдёт нормально, а 
`in = new DataInputStream(socket.getInputStream());` вывалится, кто освободит созданный выше out

maksymdobrynin replied:
Надо, если ресурс уже занят, то его разумеется надо высвободить.
В твоем примеру, надо будет закрыть, по крайне мере точно проверить не является нулом и если объект - закрываем.
Исключения в конструкторах не самая частая практика, но бывает. Чаще всегда, поток не нужен слишком долго "Поработал и закрыл", но это может отличаться от приложения к приложению, например как у нас.

Commit:
defeat: применить аксиоматическое логирование исключений валящихся из in/out стримов

Мотивация
========
Надо. Просто надо избавиться от вкючения в коде пустого блока `catch`.

>Замечание 1 так и осталось не закрытым

Чем так принципиально отличаются приведённые листинги? И там, и там - мы глотаем/давим исключение из потока ввода/вывода;
и там, и там - обрабатыавем исключение - выполнением `markClosed();`. С точки зрения выполнения программы оба этих случая
абсоолютно эквивалентны, но почему тогда первый - "плохой", а второй - "хороший" (ПОМИМО ДОГМАТИЧЕСКОЙ АКСИОМАТИКИ)?

Логика заложена такой, что закрывшийся стрим - сигнал сворачиваться, и мы не считаем это какой-то исключительной
ситуацией - просто обрабатываем исключение ответным вызовом `markClosed();` - всё. Почему по вашему мнению в одном
случае надо логировать/пробрасывать наверх, а вдругим не надо? - случаи совершенно идентичные, и их обработка идентична.
Случай №1 плох только тем, как я вижу в указанной мотивации, что у него в теле "catch" ниодной строки, хотя по результату
выполнения он совершенно идентичен случаю №2.

Ок. Допустим, мы жить не можем без логирования, или проброса - реализуем его в случае №1. Но, тогда и случай №2 надо
обрабатывать ровно тем же образом, хотя такого требования ниразу не прозвучало.

По моей логике, мы уже обработали `IOException` из соответствующего стрима и нет совершенно никакой мотивации чтобы светить
его дальше где бы то нибыло, не все исключения нуждаются в логировании. Если исключение корректно и штатно обработано, то
оно стало просто иснтрументом управления выполнением программы, рантайма, не более того.

"Замечание 1" по прежнему не принимается.

```Java
        try {
            in.readUTF();
        } catch (IOException e) {
            /**
             * А вот у нас нова возникает глухой catch.
             * Второй раз мы вызывать `markClosed()` больше не будем, да только от catch-а никуда не избавимся
             * `in.readUTF();` очень сильно наставивает, что IOException надо поймать.
             */
        } finally {
            markClosed();
        }
```

```Java
        try {
            out.writeUTF(message);
        } catch (IOException e) {
            markClosed();
        }
```

>Второй момент, если делать пустой catch, тогда происходит искусвенное подавление исключения и оно уже ничего не будет фигурировать

Именно так. М ыего умышленно давим, т.к. исключение уже корректно обработано согласно штатной логике работы программы.
А развешивать из всех углов портянки неадекватных никому ненужных стеков - действительно плохая практика.
Это как диагноз "собирательстко".

Простоты ради, и экономии времени - сверну диспут в пользу варианта "Ок". Надеюсь после этого вопросов не останется.
